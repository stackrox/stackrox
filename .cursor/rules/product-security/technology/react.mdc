---
description: Secure coding rules for React projects ⚛️
globs:
  - "**/*.jsx"
  - "**/*.tsx"
alwaysApply: false
---

- Use default React data binding to prevent XSS.
- Sanitize HTML before using dangerouslySetInnerHTML.
- Validate URLs before using them in the DOM.
- Secure server-side rendering and avoid JSON injection.
