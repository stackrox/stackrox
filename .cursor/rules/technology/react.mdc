# React-specific secure coding rules for Cursor
# This file provides security best practices for React projects.
---
description: Secure coding rules for React projects ⚛️
globs:
  - "**/*.jsx"
  - "**/*.tsx"
  - "**/src/**/*.jsx"
  - "**/src/**/*.tsx"
  - "**/components/**/*.jsx"
  - "**/components/**/*.tsx"
  - "**/pages/**/*.jsx"
  - "**/pages/**/*.tsx"
  - "**/App.jsx"
  - "**/App.tsx"
  - "**/index.jsx"
  - "**/index.tsx"
alwaysApply: false
---

- Use default React data binding to prevent XSS.
- Sanitize HTML before using dangerouslySetInnerHTML.
- Validate URLs before using them in the DOM.
- Secure server-side rendering and avoid JSON injection.
