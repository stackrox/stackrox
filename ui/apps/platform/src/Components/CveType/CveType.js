import React from 'react';
import PropTypes from 'prop-types';

const cveTypes = ['IMAGE_CVE', 'K8S_CVE', 'ISTIO_CVE', 'NODE_CVE', 'OPENSHIFT_CVE'];
const cveTypeMap = {
    IMAGE_CVE: 'Image CVE',
    K8S_CVE: 'Kubernetes CVE',
    ISTIO_CVE: 'Istio CVE',
    NODE_CVE: 'Node CVE',
    OPENSHIFT_CVE: 'OpenShift CVE',
};

const CveType = ({ types }) => {
    const sortedTypes = types.map((x) => cveTypeMap[x] || 'Unknown').sort();

    return (
        <span>
            <div className="flex flex-col">
                {sortedTypes.map((cveType) => (
                    <div className="flex justify-center" key={cveType}>
                        {cveType}
                    </div>
                ))}
            </div>
        </span>
    );
};

CveType.propTypes = {
    types: PropTypes.arrayOf(PropTypes.oneOf(cveTypes)),
};

CveType.defaultProps = {
    types: [],
};

export default CveType;
