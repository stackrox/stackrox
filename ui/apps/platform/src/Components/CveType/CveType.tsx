import React from 'react';

export type CveType =
    | 'IMAGE_CVE'
    | 'IMAGE'
    | 'IMAGE_COMPONENT'
    | 'DEPLOYMENT'
    | 'NAMESPACE'
    | 'K8S_CVE'
    | 'ISTIO_CVE'
    | 'NODE_CVE'
    | 'NODE'
    | 'NODE_COMPONENT'
    | 'OPENSHIFT_CVE';
export type CveTypeProps = {
    types: CveType[];
};

const cveTypeMap = {
    IMAGE_CVE: 'Image CVE',
    IMAGE: 'Image CVE',
    IMAGE_COMPONENT: 'Image CVE',
    DEPLOYMENT: 'Image CVE',
    NAMESPACE: 'Image CVE',
    K8S_CVE: 'Kubernetes CVE',
    ISTIO_CVE: 'Istio CVE',
    NODE_CVE: 'Node CVE',
    NODE: 'Node CVE',
    NODE_COMPONENT: 'Node CVE',
    OPENSHIFT_CVE: 'OpenShift CVE',
};

const CveType = ({ types = [] }: CveTypeProps) => {
    const sortedTypes = types.map((x) => cveTypeMap[x] || 'Unknown').sort();

    return (
        <span>
            <div className="flex flex-col">
                {sortedTypes.map((cveType) => (
                    <div className="flex justify-center" key={cveType}>
                        {cveType}
                    </div>
                ))}
            </div>
        </span>
    );
};

export default CveType;
