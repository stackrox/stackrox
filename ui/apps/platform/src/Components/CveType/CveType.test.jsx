import React from 'react';
import { render } from '@testing-library/react';

import CveType from './CveType';

describe('CveType', () => {
    describe('default style', () => {
        test('should show an image CVE type in default style', async () => {
            const { asFragment } = render(<CveType types={['IMAGE_CVE']} />);

            expect(asFragment()).toMatchSnapshot();
        });

        test('should show a node CVE type in default style', async () => {
            const { asFragment } = render(<CveType types={['NODE_CVE']} />);

            expect(asFragment()).toMatchSnapshot();
        });

        test('should show a Kubernetes CVE type in default style', async () => {
            const { asFragment } = render(<CveType types={['K8S_CVE']} />);

            expect(asFragment()).toMatchSnapshot();
        });

        test('should show an OpenShift CVE type in default style', async () => {
            const { asFragment } = render(<CveType types={['OPENSHIFT_CVE']} />);

            expect(asFragment()).toMatchSnapshot();
        });

        test('should show an Istio CVE type in default style', async () => {
            const { asFragment } = render(<CveType types={['ISTIO_CVE']} />);

            expect(asFragment()).toMatchSnapshot();
        });

        test('should show an Image and Node CVE type in default style', async () => {
            const { asFragment } = render(<CveType types={['IMAGE_CVE', 'NODE_CVE']} />);

            expect(asFragment()).toMatchSnapshot();
        });

        test('should show a Node and Kubernetes CVE type in default style', async () => {
            const { asFragment } = render(<CveType types={['NODE_CVE', 'K8S_CVE']} />);

            expect(asFragment()).toMatchSnapshot();
        });

        test('should show a Kubernetes and OpenShift CVE type in default style', async () => {
            const { asFragment } = render(<CveType types={['K8S_CVE', 'OPENSHIFT_CVE']} />);

            expect(asFragment()).toMatchSnapshot();
        });
    });
});
