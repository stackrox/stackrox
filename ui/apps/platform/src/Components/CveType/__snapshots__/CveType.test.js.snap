// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CveType default style should show a Kubernetes CVE type in default style 1`] = `
<DocumentFragment>
  <span>
    <div
      class="flex flex-col"
    >
      <div
        class="flex justify-center"
      >
        Kubernetes CVE
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`CveType default style should show a Kubernetes and OpenShift CVE type in default style 1`] = `
<DocumentFragment>
  <span>
    <div
      class="flex flex-col"
    >
      <div
        class="flex justify-center"
      >
        Kubernetes CVE
      </div>
      <div
        class="flex justify-center"
      >
        OpenShift CVE
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`CveType default style should show a Node and Kubernetes CVE type in default style 1`] = `
<DocumentFragment>
  <span>
    <div
      class="flex flex-col"
    >
      <div
        class="flex justify-center"
      >
        Kubernetes CVE
      </div>
      <div
        class="flex justify-center"
      >
        Node CVE
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`CveType default style should show a node CVE type in default style 1`] = `
<DocumentFragment>
  <span>
    <div
      class="flex flex-col"
    >
      <div
        class="flex justify-center"
      >
        Node CVE
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`CveType default style should show an Image and Node CVE type in default style 1`] = `
<DocumentFragment>
  <span>
    <div
      class="flex flex-col"
    >
      <div
        class="flex justify-center"
      >
        Image CVE
      </div>
      <div
        class="flex justify-center"
      >
        Node CVE
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`CveType default style should show an Istio CVE type in default style 1`] = `
<DocumentFragment>
  <span>
    <div
      class="flex flex-col"
    >
      <div
        class="flex justify-center"
      >
        Istio CVE
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`CveType default style should show an OpenShift CVE type in default style 1`] = `
<DocumentFragment>
  <span>
    <div
      class="flex flex-col"
    >
      <div
        class="flex justify-center"
      >
        OpenShift CVE
      </div>
    </div>
  </span>
</DocumentFragment>
`;

exports[`CveType default style should show an image CVE type in default style 1`] = `
<DocumentFragment>
  <span>
    <div
      class="flex flex-col"
    >
      <div
        class="flex justify-center"
      >
        Image CVE
      </div>
    </div>
  </span>
</DocumentFragment>
`;
